### PR TITLE
expression: rewrite builtin function: NULLIF

### PR DESCRIPTION
**Pull request by breeswish**
to https://github.com/pingcap/tidb/issues/4080

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu @winkyao @jackysp  PTAL

File: expression/expression.go [hunk 1]
```diff
@@ -33,10 +33,10 @@ import (
 // Error instances.
 var (
 	errInvalidOperation        = terror.ClassExpression.New(codeInvalidOperation, "invalid operation")
-	errIncorrectParameterCount = terror.ClassExpression.New(codeIncorrectParameterCount, "Incorrect parameter count in the call to native function '%s'")
 	errFunctionNotExists       = terror.ClassExpression.New(codeFunctionNotExists, "FUNCTION %s does not exist")
 	errZlibZData               = terror.ClassTypes.New(codeZlibZData, "ZLIB: Input data corrupted")
 	errIncorrectArgs           = terror.ClassExpression.New(codeIncorrectArgs, mysql.MySQLErrName[mysql.ErrWrongArguments])
+	ErrIncorrectParameterCount = terror.ClassExpression.New(codeIncorrectParameterCount, "Incorrect parameter count in the call to native function '%s'")
```
Comment (XuHuaiyu) [inline expression/expression.go]: why change this?

File: plan/expression_rewriter.go [hunk 2]
```diff
@@ -1030,13 +1030,46 @@ func (er *expressionRewriter) checkArgsOneColumn(args ...expression.Expression)
 	}
 }
 
+func (er *expressionRewriter) rewriteFuncCall(v *ast.FuncCallExpr) bool {
+	switch v.FnName.L {
+	case "nullif":
```
Comment (XuHuaiyu) [inline plan/expression_rewriter.go]: s/ "nullif"/ ast.NullIf ?

File: plan/expression_rewriter.go [hunk 3]
```diff
@@ -1030,13 +1030,46 @@ func (er *expressionRewriter) checkArgsOneColumn(args ...expression.Expression)
...
+		param1 := er.ctxStack[stackLen-2]
+		param2 := er.ctxStack[stackLen-1]
+		// param1 = param2
+		funcCompare, err := er.constructBinaryOpFunction(param1, param2, ast.EQ)
+		if err != nil {
+			er.err = err
+			return true
+		}
+		// if(param1 = param2, null, param1)
+		funcIf, err := expression.NewFunction(er.ctx, "if", &v.Type, funcCompare, expression.Null, param1)
```
Comment (XuHuaiyu) [inline plan/expression_rewriter.go]: s/ "if"/ ast.If

File: plan/expression_rewriter.go [hunk 4]
```diff
@@ -1030,13 +1030,46 @@ func (er *expressionRewriter) checkArgsOneColumn(args ...expression.Expression)
 	}
 }
 
+func (er *expressionRewriter) rewriteFuncCall(v *ast.FuncCallExpr) bool {
```
Comment (XuHuaiyu) [inline plan/expression_rewriter.go]: add comment for this function.

Comment (XuHuaiyu) [inline plan/expression_rewriter.go]: why return bool?

File: plan/typeinfer_test.go [hunk 5]
```diff
@@ -648,6 +648,21 @@ func (s *testPlanSuite) createTestCase4CompareFuncs() []typeInferTestCase {
 		{"isnull(c_blob     )", mysql.TypeLonglong, charset.CharsetBin, mysql.BinaryFlag, 1, 0},
 		{"isnull(c_set      )", mysql.TypeLonglong, charset.CharsetBin, mysql.BinaryFlag, 1, 0},
 		{"isnull(c_enum     )", mysql.TypeLonglong, charset.CharsetBin, mysql.BinaryFlag, 1, 0},
+
+		{"nullif(c_int      , 123)", mysql.TypeLong, charset.CharsetBin, mysql.BinaryFlag, 11, types.UnspecifiedLength},                       // TODO: tp should be TypeLonglong, decimal should be 0
+		{"nullif(c_bigint   , 123)", mysql.TypeLonglong, charset.CharsetBin, mysql.BinaryFlag, 21, types.UnspecifiedLength},                   // TODO: flen should be 20, decimal should be 0
+		{"nullif(c_float    , 123)", mysql.TypeFloat, charset.CharsetBin, mysql.BinaryFlag, types.UnspecifiedLength, types.UnspecifiedLength}, // TODO: tp should be TypeDouble
```
Comment (XuHuaiyu) [inline plan/typeinfer_test.go]: keep the gap length of the comments to code consistent.

File: expression/expression.go (hunk 1 above)
Comment (breeswish) [inline expression/expression.go]: So that plan/expression_rewriter can use this error code.

File: plan/expression_rewriter.go (hunk 4 above)
Comment (breeswish) [inline plan/expression_rewriter.go]: Indicate whether a rewrite is performed. I will add comment for this.

**Comment by XuHuaiyu** [comment]
@breeswish 
please address the comments.

File: plan/typeinfer_test.go (hunk 5 above)
Comment (breeswish) [inline plan/typeinfer_test.go]: The gap length is formatted by gofmt and we can't take control of it.

**Comment by breeswish** [comment]
@XuHuaiyu Done, PTAL

**Pull request merged**